Consignment Pricing Prediction

Data Collection

Dataset Source - https://www.kaggle.com/datasets/divyeshardeshana/supply-chain-shipment-pricing-data
The data consists of 33 columns and 10325 rows (appx.).

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

Import the CSV Data as Pandas DataFrame


In [3]:
df = pd.read_csv("dataset\SCMS_Delivery_History_Dataset_1.csv")

Show Top 5 Records


In [4]:
df.head()

ID Project Code            PQ # PO / SO # ASN/DN #        Country   
0   1   100-CI-T01  Pre-PQ Process    SCMS-4    ASN-8  Côte d'Ivoire  \
1   3   108-VN-T01  Pre-PQ Process   SCMS-13   ASN-85        Vietnam   
2   4   100-CI-T01  Pre-PQ Process   SCMS-20   ASN-14  Côte d'Ivoire   
3  15   108-VN-T01  Pre-PQ Process   SCMS-78   ASN-50        Vietnam   
4  16   108-VN-T01  Pre-PQ Process   SCMS-81   ASN-55        Vietnam   

  Managed By  Fulfill Via Vendor INCO Term Shipment Mode  ...   
0   PMO - US  Direct Drop              EXW           Air  ...  \
1   PMO - US  Direct Drop              EXW           Air  ...   
2   PMO - US  Direct Drop              FCA           Air  ...   
3   PMO - US  Direct Drop              EXW           Air  ...   
4   PMO - US  Direct Drop              EXW           Air  ...   

  Unit of Measure (Per Pack) Line Item Quantity Line Item Value Pack Price   
0                         30                 19           551.0      29.00  \
1                        240               1000          6200.0       6.20   
2                        100                500         40000.0      80.00   
3                         60              31920        127360.8       3.99   
4                         60              38000        121600.0       3.20   

  Unit Price             Manufacturing Site First Line Designation   
0       0.97     Ranbaxy Fine Chemicals LTD                    Yes  \
1       0.03      Aurobindo Unit III, India                    Yes   
2       0.80  ABBVIE GmbH & Co.KG Wiesbaden                    Yes   
3       0.07  Ranbaxy, Paonta Shahib, India                    Yes   
4       0.05      Aurobindo Unit III, India                    Yes   

  Weight (Kilograms) Freight Cost (USD) Line Item Insurance (USD)  
0                 13             780.34                       NaN  
1                358             4521.5                       NaN  
2                171            1653.78                       NaN  
3               1855           16007.06                       NaN  
4               7590           45450.08                       NaN  

[5 rows x 33 columns]

Shape of the dataset


In [6]:
df.shape

(10324, 33)

Check Missing values

In [7]:
df.isna().sum()

ID                                 0
Project Code                       0
PQ #                               0
PO / SO #                          0
ASN/DN #                           0
Country                            0
Managed By                         0
Fulfill Via                        0
Vendor INCO Term                   0
Shipment Mode                    360
PQ First Sent to Client Date       0
PO Sent to Vendor Date             0
Scheduled Delivery Date            0
Delivered to Client Date           0
Delivery Recorded Date             0
Product Group                      0
Sub Classification                 0
Vendor                             0
Item Description                   0
Molecule/Test Type                 0
Brand                              0
Dosage                          1736
Dosage Form                        0
Unit of Measure (Per Pack)         0
Line Item Quantity                 0
Line Item Value                    0
Pack Price                         0
U

In [8]:
df.duplicated().sum()


0

In [9]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10324 entries, 0 to 10323
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ID                            10324 non-null  int64  
 1   Project Code                  10324 non-null  object 
 2   PQ #                          10324 non-null  object 
 3   PO / SO #                     10324 non-null  object 
 4   ASN/DN #                      10324 non-null  object 
 5   Country                       10324 non-null  object 
 6   Managed By                    10324 non-null  object 
 7   Fulfill Via                   10324 non-null  object 
 8   Vendor INCO Term              10324 non-null  object 
 9   Shipment Mode                 9964 non-null   object 
 10  PQ First Sent to Client Date  10324 non-null  object 
 11  PO Sent to Vendor Date        10324 non-null  object 
 12  Scheduled Delivery Date       10324 non-null  object 
 13  D

In [10]:
df.nunique()


ID                              10324
Project Code                      142
PQ #                             1237
PO / SO #                        6233
ASN/DN #                         7030
Country                            43
Managed By                          4
Fulfill Via                         2
Vendor INCO Term                    8
Shipment Mode                       4
PQ First Sent to Client Date      765
PO Sent to Vendor Date            897
Scheduled Delivery Date          2006
Delivered to Client Date         2093
Delivery Recorded Date           2042
Product Group                       5
Sub Classification                  6
Vendor                             73
Item Description                  184
Molecule/Test Type                 86
Brand                              48
Dosage                             54
Dosage Form                        17
Unit of Measure (Per Pack)         31
Line Item Quantity               5065
Line Item Value                  8741
Pack Price  

In [11]:
df.describe()


ID  Unit of Measure (Per Pack)  Line Item Quantity   
count  10324.000000                10324.000000        10324.000000  \
mean   51098.968229                   77.990895        18332.534870   
std    31944.332496                   76.579764        40035.302961   
min        1.000000                    1.000000            1.000000   
25%    12795.750000                   30.000000          408.000000   
50%    57540.500000                   60.000000         3000.000000   
75%    83648.250000                   90.000000        17039.750000   
max    86823.000000                 1000.000000       619999.000000   

       Line Item Value    Pack Price    Unit Price  Line Item Insurance (USD)  
count     1.032400e+04  10324.000000  10324.000000               10037.000000  
mean      1.576506e+05     21.910241      0.611701                 240.117626  
std       3.452921e+05     45.609223      3.275808                 500.190568  
min       0.000000e+00      0.000000      0.000000                   0.000000  
25%       4.314593e+03      4.120000      0.080000                   6.510000  
50%       3.047147e+04      9.300000      0.160000                  47.040000  
75%       1.664471e+05     23.592500      0.470000                 252.400000  
max       5.951990e+06   1345.640000    238.650000                7708.440000

#Exploring Data


In [14]:
df.head()

ID Project Code            PQ # PO / SO # ASN/DN #        Country   
0   1   100-CI-T01  Pre-PQ Process    SCMS-4    ASN-8  Côte d'Ivoire  \
1   3   108-VN-T01  Pre-PQ Process   SCMS-13   ASN-85        Vietnam   
2   4   100-CI-T01  Pre-PQ Process   SCMS-20   ASN-14  Côte d'Ivoire   
3  15   108-VN-T01  Pre-PQ Process   SCMS-78   ASN-50        Vietnam   
4  16   108-VN-T01  Pre-PQ Process   SCMS-81   ASN-55        Vietnam   

  Managed By  Fulfill Via Vendor INCO Term Shipment Mode  ...   
0   PMO - US  Direct Drop              EXW           Air  ...  \
1   PMO - US  Direct Drop              EXW           Air  ...   
2   PMO - US  Direct Drop              FCA           Air  ...   
3   PMO - US  Direct Drop              EXW           Air  ...   
4   PMO - US  Direct Drop              EXW           Air  ...   

  Unit of Measure (Per Pack) Line Item Quantity Line Item Value Pack Price   
0                         30                 19           551.0      29.00  \
1                        240               1000          6200.0       6.20   
2                        100                500         40000.0      80.00   
3                         60              31920        127360.8       3.99   
4                         60              38000        121600.0       3.20   

  Unit Price             Manufacturing Site First Line Designation   
0       0.97     Ranbaxy Fine Chemicals LTD                    Yes  \
1       0.03      Aurobindo Unit III, India                    Yes   
2       0.80  ABBVIE GmbH & Co.KG Wiesbaden                    Yes   
3       0.07  Ranbaxy, Paonta Shahib, India                    Yes   
4       0.05      Aurobindo Unit III, India                    Yes   

  Weight (Kilograms) Freight Cost (USD) Line Item Insurance (USD)  
0                 13             780.34                       NaN  
1                358             4521.5                       NaN  
2                171            1653.78                       NaN  
3               1855           16007.06                       NaN  
4               7590           45450.08                       NaN  

[5 rows x 33 columns]

FREIGHT COST ANALYSIS:


1. https://truckstop.com/blog/freight-shipping-costs/
2. https://www.gojarrett.com/blog/key-factors-that-influence-your-freight-costs
From the above links, the factors that affect freight costs are:


Distance: The farther the destination, the more it will cost to ship the goods.


Weight: Heavier items will require more fuel to transport and may incur additional handling fees.


Size: Large or bulky items may require special handling or equipment, which can increase the cost of shipping.


Mode of transportation: The method of transport used (e.g., air, ocean, truck, rail) will affect the cost.


Carrier: Different carriers have different pricing structures and fees, so the choice of carrier can affect the cost of shipping.


Destination: Shipping to remote or hard-to-reach locations may incur additional fees.


Packaging: Properly packaging items to prevent damage during transport can reduce the risk of additional fees.


Type of goods: Hazardous or fragile items may require special handling, which can increase the cost of shipping.



From the above knowledge, a total of 11 columns from the data set are selected, which are:


      1. Country 2. Shipment Mode 3. Unit of measure 4. Line Item Quantity 5. Line Item Value 6. Pack price 7. Unit Price 8. Weight (kilos) 9. Freight Cost (USD) [For Prediction] 10. Vendor INCO Term 11. Fulfil via

In [39]:
#Remove columns that are not required for prediction
list_not_required_coulumns = ['ID', 'Project Code', 'PQ #', 'PO / SO #', 'ASN/DN #',
       'Managed By','PQ First Sent to Client Date', 'PO Sent to Vendor Date',
       'Scheduled Delivery Date', 'Delivered to Client Date',
       'Delivery Recorded Date', 'Product Group', 'Sub Classification',
       'Vendor', 'Item Description', 'Molecule/Test Type', 'Brand', 'Dosage',
       'Dosage Form','Manufacturing Site',
       'First Line Designation',
       'Line Item Insurance (USD)']
df = df.drop(list_not_required_coulumns,axis = 1)
df.head()

Country  Fulfill Via Vendor INCO Term Shipment Mode   
0  Côte d'Ivoire  Direct Drop              EXW           Air  \
1        Vietnam  Direct Drop              EXW           Air   
2  Côte d'Ivoire  Direct Drop              FCA           Air   
3        Vietnam  Direct Drop              EXW           Air   
4        Vietnam  Direct Drop              EXW           Air   

   Unit of Measure (Per Pack)  Line Item Quantity  Line Item Value   
0                          30                  19            551.0  \
1                         240                1000           6200.0   
2                         100                 500          40000.0   
3                          60               31920         127360.8   
4                          60               38000         121600.0   

   Pack Price  Unit Price Weight (Kilograms) Freight Cost (USD)  
0       29.00        0.97                 13             780.34  
1        6.20        0.03                358             4521.5  
2       80.00        0.80                171            1653.78  
3        3.99        0.07               1855           16007.06  
4        3.20        0.05               7590           45450.08

In [43]:
#Rename columns
df = df.rename(columns={"Country":"country","Fulfill Via":"fulfill_via","Vendor INCO Term":"vendor",
                   "Shipment Mode":"shipment_mode", "Unit of Measure (Per Pack)":"unit_of_measure",
                   "Line Item Quantity":"line_item_quantity","Pack Price":"pack_price",
                   "Unit Price":"unit_price","Weight (Kilograms)":"weight","Freight Cost (USD)":"freight_cost",
                   "Line Item Value":"line_item_value"})

In [44]:
df.head(2)

country  fulfill_via vendor shipment_mode  unit_of_measure   
0  Côte d'Ivoire  Direct Drop    EXW           Air               30  \
1        Vietnam  Direct Drop    EXW           Air              240   

   line_item_quantity  line_item_value  pack_price  unit_price weight   
0                  19            551.0        29.0        0.97     13  \
1                1000           6200.0         6.2        0.03    358   

  freight_cost  
0       780.34  
1       4521.5

In [45]:
print("categories in country ",end=" ")
print(df["country"].unique())

categories in country  ["Côte d'Ivoire" 'Vietnam' 'Nigeria' 'Zambia' 'Tanzania' 'Rwanda' 'Haiti'
 'Zimbabwe' 'Ethiopia' 'South Africa' 'Guyana' 'Namibia' 'Botswana'
 'Mozambique' 'Kenya' 'Kazakhstan' 'Uganda' 'Kyrgyzstan' 'Senegal' 'Benin'
 'Lesotho' 'Pakistan' 'Swaziland' 'Ghana' 'Angola' 'Lebanon'
 'Sierra Leone' 'Cameroon' 'South Sudan' 'Burundi' 'Dominican Republic'
 'Malawi' 'Congo, DRC' 'Sudan' 'Mali' 'Guatemala' 'Togo' 'Afghanistan'
 'Liberia' 'Burkina Faso' 'Guinea' 'Libya' 'Belize']


In [47]:
d_country = {}
for i in df["country"]:
    if i in d_country:
        d_country[i] = d_country[i] + 1
    else:
        d_country[i] = 1
sorted_dict = sorted(d_country.items(),key = lambda kv: kv[1],reverse=True)
len_total_countries =int(len(d_country) )
sum_total_countries = int(sum(d_country.values()))
avg = sum_total_countries/len_total_countries
print(avg)
print(sorted_dict)

240.09302325581396
[('South Africa', 1406), ('Nigeria', 1194), ("Côte d'Ivoire", 1083), ('Uganda', 779), ('Vietnam', 688), ('Zambia', 683), ('Haiti', 655), ('Mozambique', 631), ('Zimbabwe', 538), ('Tanzania', 519), ('Rwanda', 430), ('Congo, DRC', 333), ('Guyana', 237), ('Ethiopia', 216), ('South Sudan', 164), ('Kenya', 111), ('Burundi', 98), ('Namibia', 95), ('Cameroon', 75), ('Botswana', 70), ('Ghana', 58), ('Dominican Republic', 52), ('Sudan', 46), ('Swaziland', 35), ('Mali', 17), ('Pakistan', 15), ('Guatemala', 15), ('Malawi', 14), ('Benin', 13), ('Lebanon', 8), ('Libya', 8), ('Angola', 7), ('Liberia', 6), ('Lesotho', 4), ('Sierra Leone', 4), ('Senegal', 3), ('Togo', 3), ('Afghanistan', 3), ('Kazakhstan', 2), ('Kyrgyzstan', 2), ('Burkina Faso', 2), ('Guinea', 1), ('Belize', 1)]


df["Country] = From the above calculation, the average of all country counts is about 240. For the machine learning model, the countries that are selected are above count 240.

In [ ]:
ls = [ 'Ethiopia' , 'Guyana' 'Namibia' 'Botswana'
  'Kenya' 'Kazakhstan' 'Kyrgyzstan' 'Senegal' 'Benin'
 'Lesotho' 'Pakistan' 'Swaziland' 'Ghana' 'Angola' 'Lebanon'
 'Sierra Leone' 'Cameroon' 'South Sudan' 'Burundi' 'Dominican Republic'
 'Malawi' 'Sudan' 'Mali' 'Guatemala' 'Togo' 'Afghanistan'
 'Liberia' 'Burkina Faso' 'Guinea' 'Libya' 'Belize']